suite/tests/api/ir_aarch64.c

```diff
@@ -3960,6 +3960,71 @@ test_asimddiff(void *dc)
     test_instr_encoding(dc, OP_umull2, instr);
 }
 
+static void
+test_sys(void *dc)
+{
+    byte *pc;
+    instr_t *instr;
+
+    /* SYS #<op1>, <Cn>, <Cm>, #<op2>{, <Xt>}
+     *
+     * Data cache operations are aliases of SYS:
+     * DC <dc_op>, <Xt>
+     * is equivalent to
+     * SYS #<op1>, C7, <Cm>, #<op2>, <Xt>
+     */
+
+    /* DC ZVA, Xt => SYS #3, C7, C4, #1, Xt */
```

Comment:
Also, on the analysis side: for something like #4328, drcachesim needs to identify cache flush instructions.  That would similarly be easier without checking hardcoded numeric values for 4 different immediates.  Some kind of separate opcode, or at least named constants, would be nicer.

This is somewhat similar, though more significant, to the prefetch type distinction, which @abhinav92003 recently put in and had to do decoding of immediates in drcachesim: https://github.com/DynamoRIO/dynamorio/blob/master/clients/drcachesim/tracer/instru.cpp#L82.  IMHO it would be nicer to have DR provide such decoding and not force clients to look at immediate bits.  It could again be separate opcodes or at least named constants (xref DR_DMB_OSHLD, etc. where arm/insr_create.h has named constants for `OP_dmb`).

Reply:
Note that there is precedent in DR for splitting opcodes where convenient for tools and not strictly matching ISA manuals.  Assembler mnemonics are using `DC ZVA` and not `SYS 0xxx 0xx 0xxx 0xxx` anyway so separate opcodes need not be seen as divergent from standard practices; though separating only in INSTR_CREATE and providing named constants is another way to go.

Reply:
Good point about divergence!

Just to be clear, do you mean having e.g. `OP_dc_zva` as an opcode rather than having `OP_sys_dc` as an opcode with constructors like `INSTR_CREATE_dc_zva` adding ZVA as operands?

Reply:
I'm not sure what the best choice is.  I would definitely vote for separating out `dc` from `sys`: I think there are few downsides and lots of benefits there.  For separating further in the opcodes: `OP_dc_zva`, `OP_dc_cvac`, etc.: I kind of like that as it seems to match other ISA's better, with each opcode a distinct semantic operation.  But then I'd want to also split `OP_prfm` up, and that also brings up `OP_dmb`.  If there were just one "sub-opcode" enum of named constants, like for `OP_dmb` with `DR_DMB_OSHLD`, etc., that might be a good compromise?  So `OP_dc` with `DR_DC_ZVA`, etc., with a client dispatching on a single immediate operand?  The creation would then be `INSTR_CREATE_dc` taking in a single immediate.  But completely separate opcodes doesn't seem so bad either.

Reply:
Maybe get more opinions?  @johnfxgalea has written a bunch of clients; so has @zhaoqin in the past at least.

Reply:
The tricky bit here is that DC ZVA is a bit of an odd one out, as it is not really a cache maintentance operation (but a store zero operation). Perhaps you'd want some OP_dc_flush and OP_dc_zva then to separate between the two categories?

Reply:
Do not know much about details or context here, just my two cents here.

IMHO, the opcode should be the semantic behavior of the instruction, and the operand should be the affected target, e.g., data type: memory, reg, immed, data size: 1 byte, 2 bytes etc.

Instruction like OP_dmb with the DR_DMB_* enum is a bit strange to me. Those enums are more like a semantic policy.
It is like we have an OP_shift instruction and the operand is left/right/arith/logic.

So I would prefer separate opcode for each semantic behavior.

Reply:
Personally, I would keep the IR as close to the ISA as possible. If A64 sees DC as a separate single opcode, then I think DR should do the same. For instance, on Intel, there is the VPCMPW instruction. This instruction takes an imm operand which defines how the mask should be constructed, e.g., if the value is 0, do an equality comparison, if the value is 1, do a less then comparison. In this case, DynamoRIO does not have a separate opcode for each case.

Reply:
DC, as well as AT, CFP, CPP, DVP, IC and TLBI are all aliases of the SYS instruction at specific indexes that the SYS operates on. The Arm ARM states that the use of the alias mnemonics is preferred for SYS instructions at those indexes. However, that would still give you a single 'DC' instruction, and not a way to distinguish between DC ZVA and the other DC flush operations.

Reply:
Oh, I am in agreement with separating DC from SYS. I am talking about whether further opcodes like OP_dc_zva and OP_dc_cvac should be added.

Don't hold a strong opinion, just my two cents :p

Reply:
In the past we generally followed what @zhaoqin and I are espousing about the opcode defining semantics and the operands defining only the degree/scale/size and which registers/addresses.  IMHO that makes it easier to write analysis code. For 32-bit ARM, we did things like splitting up `VST1` into `OP_vst1_{8,16,32,64,lane_8,lane_16,lane_32}`.  Looks like we weren't 100% consistent though: we did not split up `DMB`, following the ISA there with a semantic immediate left exposed in the IR, probably b/c we just didn't consider it closely in the initial port.

Re: vpcmpw: looks like it's AVX-512 which brought in a ton of complex opcodes.  I think it was more that it was easiest to just add the ISA opcodes and that the semantics were not closely examined (many of those SIMD ops are complex), than that there was any conscious choice to not split it up.   I was not aware of what that opcode did until it was pointed out here.

So I guess it's your decision @AssadHashmi.  We have votes for each of these:

1. `OP_dc` with an immediate (and enum `DR_DC_ZVA`, `DR_DC_CVAC` or somesuch)
2. `OP_dc_zva`, plus `OP_dc_flush` with an immediate + enum
3. `OP_dc_zva`, `OP_dc_cvac`, etc.

Reply:
Initially I was in agreement with @johnfxgalea in that the IR should remain faithful to the ISA as far as possible and actually started doing that change yesterday. Having slept on it though, I can see that `OP_dc_zva`, `OP_dc_cvac`, etc are an improvement in terms of usability.

In this case we have a `SYS` instruction with 5 operands which has aliases in instructions like `DC` and `IC` which have 2 operands 1 of which actually tells the instruction what to do. So, 4 of `SYS`'s operands are permutations which map to a particular behaviour of `DC`, `IC` etc. To me, that justifies having OP_dc_zva, OP_dc_cvac, etc as opcodes in their own right. **In this case**.

Reply:
SGTM.  Wondering what @johnfxgalea thinks about the drcachesim code for `OP_prfm` where right now we have just the ISA opcode with immediates the client has to go and decode: https://github.com/DynamoRIO/dynamorio/blob/master/clients/drcachesim/tracer/instru.cpp#L82.  I'm not a fan of a client having to do that: it feels like DR only decoded it half-way.  Splitting the opcode would make 18 separate `OP_prfm_read_l1`, `OP_prfm_read_l1_nt`, etc. and the client has to check for it being in a range to figure out whether it's a `PRFM`, but a range check is simpler than decoding encoded bits.  I suppose leaving the single `OP_prfm` and providing a named enum where `DR_PRFM_READ_L1_NT`, etc. are set to equal their encoded 5-bit immediate value avoids the bit decoding as well.

Reply:
I agree that adding opcodes based on semantics is way better for clients. I only raised that issue b/c I want the IR to be consistent. For instance, another example relates to pshuf instructions... In an ideal world, are we going to have a different opcode for each shuffle combination?

Reply:
Also what if we have separate opcodes and the user just wants to know if an instruction is OP_prfm (regardless of the semantics). Is DR going to have helper functions that take instructions, retrieves the opcodes, and tests those opcodes against each sub (semantic-aware) opcode?

Reply:
Right, there is a tradeoff.  Yes, for the split opcode grouping into the ISA opcode, a helper would be good.  So either way there's a helper: either split opcodes with a helper to group, or a single opcode with a helper (maybe just enum for source operand integer) to obtain the sub-opcode, and there's a judgement call on where to draw the line on semantic difference consistency: prefetch types vs cache management types.  Anyway, to move this PR forward, I filed a separate #4388 on `PRFM`.

Reply:
Folks, sadly our discussion turns out to be academic!

The AArch64 codec is not able to generate decode logic for multiple encodings with the same opcode. This means the policy of IR faithfully following the ISA is **built in** to the implementation.

So, not only can we **not** have `OP_dc_zve()` etc but we can't even have `OP_dc()` !

I've just pushed the latest change :-/

Reply:
> The AArch64 codec is not able to generate decode logic for multiple encodings with the same opcode.

Not sure I follow: aren't we trying to have separate opcodes for separate encodings?  You mean there are some `SYS` encodings that overlap with the `DC` ones we're separating in some way the codec generator doesn't support today?  We can change the codec or the generator code.